gui: Add Open External Wallet action

promag opened the pull request:
This PR adds the ability to open external wallets on the GUI.

<img width="440" alt="Screenshot 2019-09-08 at 18 14 36" src="https://user-images.githubusercontent.com/3534524/64491830-a34ff680-d264-11e9-9b91-180909564d18.png">
<img width="469" alt="Screenshot 2019-09-08 at 18 13 17" src="https://user-images.githubusercontent.com/3534524/64491831-a34ff680-d264-11e9-8d37-b25a30cb9180.png">

Sjors commented:
Concept ACK, but let's put this in the Open Wallet sub menu. You could put a separator underneath the list of known wallets and then call the menu item "Open file...". I don't like the word "External" because it could get confusing if we add (e.g.) hardware wallet support.

promag commented:
@Sjors yes that was already brought up by @jnewbery and in IRC I suggested "File -> Open Wallet -> Other..." On the end of the menu.

promag commented:
@Sjors how about:
<img width="481" alt="screenshot 2019-01-19 at 17 26 36" src="https://user-images.githubusercontent.com/3534524/51430274-bc8e2f00-1c10-11e9-82b9-c736d5d43ef6.png">

jnewbery commented:
> @Sjors how about:

Yes, this looks ideal to me

promag commented:
Updated.

promag commented:
Updated, still rebased on #15195.

Sjors commented:
tACK b537d29, but note that you can only open directory based wallets

promag commented:
@Sjors right, and it looks like a pain to try to open both with Qt 😕

promag commented:
Is this for 0.18?

Sjors commented:
Needs rebase because Open Wallet got merged. That shouldn't impact my earlier tACK, so we might be able to add this to 0.18 today.

jonasschnelli commented:
Needs rebase

fanquake commented:
@promag Given that this is tagged for `0.19.0`, can you rebase?

promag commented:
Sure, in a couple of hours.

promag committed:
gui: Add OpenExternalWalletActivity

promag commented:
Rebased, and refactored. Reverted to add the action in the File menu, not in the wallets menu, so at the moment it's like:

<img width="504" alt="Screenshot 2019-09-08 at 12 57 41" src="https://user-images.githubusercontent.com/3534524/64487879-4dfdf000-d238-11e9-983d-a9d0040c64cd.png">

Previous version:

<img width="481" alt="screenshot 2019-01-19 at 17 26 36" src="https://user-images.githubusercontent.com/3534524/51430274-bc8e2f00-1c10-11e9-82b9-c736d5d43ef6.png">

@jonasschnelli @Sjors @fanquake  Let me know your thoughts.

I'll update OP after this decision.

fanquake commented:
@promag I prefer the previous version, with the action in the wallet menu. Will let other weigh in.

promag commented:
You mean the open menu with:
 - no wallets available
 - separator
 - other...

fanquake commented:
The [one currently in the PR description](https://user-images.githubusercontent.com/3534524/51430274-bc8e2f00-1c10-11e9-82b9-c736d5d43ef6.png) and [in this comment](https://github.com/bitcoin/bitcoin/pull/15204#issuecomment-455799571).

jnewbery commented:
I agree with @fanquake that keeping it in the 'Open Wallet' submenu makes more sense.

promag commented:
Thanks, also updated OP.

meshcollider commented:
Concept ACK

jonatack commented:
Concept ACK, will review this weekend.

instagibbs commented:
got this message on stderr on qt startup with this PR:
`Gtk-Message: 16:07:25.804: GtkDialog mapped without a transient parent. This is discouraged.`

I tried opening up a random wallet filed I had, but it was greyed out in the file selection? Am I missing something?

promag commented:
@instaggibbs see https://github.com/bitcoin/bitcoin/pull/15204#issuecomment-463132136

instagibbs commented:
ok that's going to have to be communicated in some obvious way...

instagibbs commented:
what does `directory based wallets` even mean? how is this detected?

ryanofsky commented:
> what does `directory based wallets` even mean i.e., how is this detected?

The underlying wallet code can only open:

1) Wallet files named wallet.dat
2) Wallet files with names other than `wallet.dat` that are located directly in the top level wallet_dir

It hasn't been possible since #11687 to even create wallet files named anything other than `wallet.dat` so the second option just exists for backwards compatibility.

Only creating wallet files named `wallet.dat` and referring to wallets by directory rather than file name was done to address concerns from luke and others https://github.com/bitcoin/bitcoin/pull/11687#issuecomment-344381289 that if users only backed up individual wallet files instead of whole wallet directories they could lose data.

ryanofsky commented:
Haven't tested this but looking at the implementation it seems like this might be lacking the error checking from the `loadwallet` RPC that ensures that the wallet which the user is trying to open actually exists, and that CreateWalletFromFIle code will not go and create at new empty wallet at the specified location:

https://github.com/bitcoin/bitcoin/blob/fb4f5beb6ede4aadeaff779cd67a0f6665419488/src/wallet/rpcwallet.cpp#L2577-L2585

Could be a pretty scary user experience to try to load an existing wallet but choose the wrong path, and see a new completely empty wallet instead of a "location does not contain wallet.dat" error.

If adding the check is necessary, I'd suggest moving the code out of the loadwallet RPC to a common location instead of adding duplicate code for the gui.

promag commented:
> Could be a pretty scary user experience to try to load an existing wallet but choose the wrong path, and see a new completely empty wallet instead of a "location does not contain wallet.dat" error.

@ryanofsky indeed, currently it's like "load or create" :trollface: will change accordingly.

> ok that's going to have to be communicated in some obvious way...

@instagibbs you mean in the release notes or do you have something else in mind?

promag committed:
refactor: Add LoadExistingWallet

instagibbs commented:
oh.... you have to click "open" while in the right directory or after clicking once on that directory holding the wallet file?

that's not even close to intuitive :grimacing:

instagibbs commented:
So this either needs an extensive explanation before the user ventures to use it, or it should allow the user to click and open a wallet.dat in a directory, in addition to the directory itself. Latter seems to make most sense?

promag commented:
@instagibbs mind testing again? Now you can select the wallet.dat or the parent directory.

@ryanofsky now it doesn't create the wallet, thanks!

Sjors commented:
On macOS, when I select a directory ~/temp/A it complains:

<img width="427" alt="Schermafbeelding 2019-09-16 om 10 45 08" src="https://user-images.githubusercontent.com/10217/64944730-4a3d1f80-d86f-11e9-9d7e-0a1c007d5273.png">

When I double click on the directory A (so it shows the wallet.dat file, etc) it still gives the above error.

It only works when I select the wallet.dat file.

MarcoFalke commented:
This has missed the deadline in #15940

jnewbery commented:
I don't think the wallet name should display the path to that wallet:

![Screenshot from 2019-09-16 11-54-07](https://user-images.githubusercontent.com/1063656/64973301-c851f400-d878-11e9-9d73-c97b76d94fa8.png)
![Screenshot from 2019-09-16 11-54-27](https://user-images.githubusercontent.com/1063656/64973302-c851f400-d878-11e9-941f-c20d2464583f.png)

ryanofsky commented:
> So this either needs an extensive explanation before the user ventures to use it, or it should allow the user to click and open a wallet.dat in a directory, in addition to the directory itself. Latter seems to make most sense?

Allowing opening of wallet.dat files seems good. Would just need to strip the last `wallet.dat` component to make the path a wallet location. Another option would be to use a directory selector instead of file selector. https://stackoverflow.com/questions/9618381/how-to-specify-the-qfiledialoggetexistingdirectory-method

> I don't think the wallet name should display the path to that wallet:

I kind of do, but anything in particular you'd like to replace it with? It should be ok to replace it with something something shorter in the GUI selector. I'd be more wary of changing the output of `getwalletinfo` RPC, or changing the invariant that whatever string you pass to `-wallet` or `loadwallet` or `createwallet` will uniquely identify the wallet while it is open, and be the wallet endpoint and `-rpcwallet` argument you pass to bitcoin-cli.

promag commented:
> Would just need to strip the last wallet.dat component to make the path a wallet location 

That's done here: https://github.com/bitcoin/bitcoin/pull/15204/commits/9e1fdd7f94161ecfbe7a4247fb2a672d5190ff83#diff-dc8c2957962a3032071c55c70e120932R331 but apparently shouldn't be done for directories.

> I don't think the wallet name should display the path to that wallet:

@jnewbery this wasn't changed here, it's already the case if you use `loadwallet` with a path. Let's move that discussion elsewhere? 

@Sjors I think I've fixed it, mind checking?

promag commented:
BTW I'm avoiding `QFileDialog` static methods, such as [QFileDialog::getOpenFileName](https://doc.qt.io/qt-5/qfiledialog.html#getOpenFileName), which have the following drawback (from the doc):
> Warning: Do not delete parent during the execution of the dialog. If you want to do this, you should create the dialog yourself using one of the QFileDialog constructors.

#15310 is related to this warning.

Sjors commented:
In 91d6e0b I'm able to to open a wallet by just selecting the directory. Opening the directory and opening the wallet.dat file work as well.

Agree with @jnewbery about using the wallet name instead of the full path. Since the RPC has the same problem, we can deal with that in a followup.

instagibbs commented:
ACK on punting the full path issue until a followup

instagibbs commented:
Now when I use it, I can only open the wallet by selecting a file within a directory with a `wallet.dat` present. It lets me "open" any file in that directory, but it opens up the specific `wallet.dat` that is in the same directory, I think.

promag commented:
I'm tempted to allow selecting either wallet.dat or a directory.

ryanofsky commented:
> I'm tempted to allow selecting either wallet.dat or a directory.

If I were implementing this, I'd use a folder selector (https://stackoverflow.com/questions/13299283/folder-browser-dialog-in-qt ), not a file selector, and only allow choosing directories containing wallet.dat files (disable the choose / select button for other directories).

But if there's a reason to list files as well as directories, I'd think there should be a way to filter the file list to only show wallet.dat files.

instagibbs commented:
> I'd use a folder selector

concept ACK on this suggestion. I don't think I would have been overly confused with this originally.

promag committed:
gui: Add Open External Wallet action

promag commented:
Done, now can only select folders.

instagibbs commented:
ok, this is much more natural. quick tACK https://github.com/bitcoin/bitcoin/pull/15204/commits/4e8cb3fd5efe2d108d8cf4017d197fb3e7c8eb40

Sjors commented:
Agree the behavior in 4e8cb3f is nicer. I haven't reviewed the code.

jonatack reviewed (commented):
Light code review and quick manual test ACK  4e8cb3fd5efe2d108d8cf4017d197fb3e7c8eb40 apart from my comments below and a bit more manual testing -- I may have encountered weird behavior as described in my comments below. I agree with @jnewbery on the full path behavior and  understand @ryanofsky's points, and don't have a proposal yet that is more concrete than having a way to see the full path that is discreet and user-friendly. Punting for a follow-up duly noted.

jonatack commented:
> I don't like the word "External" because it could get confusing if we add (e.g.) hardware wallet support.

I'm hesitant as well with the ambiguity of the word "external" -- should we be more precise?

instagibbs commented:
How about "wallet in external directory" and variations thereof

jonatack commented:
Yes, "external directory" or "in another directory" seems already better (unless the plan is to include external drives and HWs in the same category later on)

ryanofsky commented:
> How about "wallet in external directory" and variations thereof

I like the "Open a wallet in an external directory" Jon suggested

michaelfolkson commented:
The multiple wallets should be in separate directories by default? In which case why not just "Other wallet" or "Additional wallet"? I don't think attention should be brought to it being in an external directory if that is the default.

ryanofsky commented:
> The multiple wallets should be in separate directories by default? In which case why not just "Other wallet" or "Additional wallet"? I don't think attention should be brought to it being in an external directory if that is the default.

External means outside the bitcoin wallet dir (`~/.bitcoin/wallets` on linux) where the list of wallets comes from. I like your "Open other wallet..." suggestion, though. Maybe saying "external" just adds confusion.

jonatack commented:
Maybe "Open a wallet located in a different directory"

promag commented on src/qt/bitcoingui.cpp:
```diff
@@ -337,6 +337,9 @@ void BitcoinGUI::createActions()
     m_open_wallet_menu = new QMenu(this);
 
+    m_open_external_wallet_action = new QAction(tr("Other..."), this);
+    m_open_external_wallet_action->setStatusTip(tr("Open a wallet on external location"));
```
Picked "Open a wallet in an external directory".

promag commented on src/qt/walletcontroller.cpp:
```diff
@@ -281,10 +283,63 @@ void OpenWalletActivity::open(const std::string& path)
+void OpenExternalWalletActivity::open()
+{
+    Q_ASSERT(!m_file_dialog);
+    m_file_dialog = new QFileDialog(m_parent_widget, tr("Open External wallet"), QDir::homePath());
```
Changed to "Open Wallet", I think it's enough.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -157,9 +157,31 @@ std::shared_ptr<CWallet> LoadWallet(interfaces::Chain& chain, const WalletLocati
+        if (fs::symlink_status(wallet_dat_file).type() == fs::file_not_found) {
+            exists = false;
+            error = "Directory " + location.GetName() + " does not contain a wallet.dat file.";
+            return nullptr;
```
Can't reproduce, please let me know if you hit it again.

jonatack commented:
Will test more and re-review. Some edge case test ideas from today's PR review meeting:
- no wallet in directory
- multiple wallets in directory
- a non-wallet file named wallet.dat
- deleting the wallet file while trying to load it

promag commented:
> Will test more and re-review. Some edge case test ideas from today's PR review meeting:
> 
> * no wallet in directory

it already fails:

<img width="532" alt="Screenshot 2019-09-18 at 22 50 05" src="https://user-images.githubusercontent.com/3534524/65188604-b3d34e80-da66-11e9-8c64-3188da5a7a47.png">


> * multiple wallets in directory

not supported - it's already the case with RPC `loadwallet`:

```
bitcoin-cli -regtest loadwallet /Users/joao/Desktop/test.dat

Wallet file verification failed: Invalid -wallet path '/Users/joao/Desktop/test.dat'. -wallet path should point to a directory where wallet.dat and database/log.?????????? files can be stored, a location where such a directory could be created, or (for backwards compatibility) the name of an existing data file in -walletdir ("/Users/joao/Library/Application Support/Bitcoin/regtest/wallets") (code -4)
```

> * a non-wallet file named wallet.dat

not supported ☝️ 

> * deleting the wallet file while trying to load it

Didn't test this here, and not sure how to reliable test it.

fanquake commented:
re-Concept ACK. Here's the dialog on macOS:

![open_dialog](https://user-images.githubusercontent.com/863730/65215446-1b79b000-dae0-11e9-820a-41997c3f95a4.png)

A few thoughts:

It'd be good if the GUI realised when we opened a wallet that's actually inside our wallet directory via the `Other...` menu, otherwise you end up with the wallet loaded, but not greyed out:
![open](https://user-images.githubusercontent.com/863730/65210656-7f937880-dace-11e9-9408-1d7d591057a4.png)

This seems to have created at least one path to a segfault, where if you load duplicate (copy of a) wallets we don't catch the BerkeleyBatch exception. I don't think it was possible to do this via the GUI before (it's also currently caught when loading wallets via the RPC etc).
```
libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: BerkeleyBatch: Can't open database wallet.dat (duplicates fileid ccd6a006040000010ccf2f5cc535010000000000 from wallet.dat)

2019-09-19T03:02:31Z GUI: Qt has caught an exception thrown from an event handler. Throwing
exceptions from an event handler is not supported in Qt.
You must not let any exception whatsoever propagate through Qt code.
If that is not possible, in Qt 5 you must at least reimplement
QCoreApplication::notify() and catch all exceptions there.
```

> Maybe saying "external" just adds confusion.

Agree.

Noticed at least one quirk where Qt fails to display file paths correctly:

<img width="972" alt="path_confusion" src="https://user-images.githubusercontent.com/863730/65215478-3e0bc900-dae0-11e9-955a-a0ed38a13877.png">

promag commented:
> It'd be good if the GUI realised when we opened a wallet that's actually inside our wallet directory via the `Other...` menu, otherwise you end up with the wallet loaded, but not greyed out:

Yes, different PR I guess.

jonatack commented:
@promag Thanks for the update. While testing f60f80e599284b4f098b54766e2221ebcd11f587, I reproduced the [issue mentioned above](https://github.com/bitcoin/bitcoin/pull/15204/#discussion_r325757097) on  Debian 4.19.37-5+deb10u2 (2019-08-08) x86_64 GNU/Linux. Opening a folder causes the GUI display the error, then a second later attempts to open the wallet anyway and hangs in this state.

![Screenshot from 2019-09-19 12-26-00](https://user-images.githubusercontent.com/2415484/65236933-21b06200-dac9-11e9-9268-21695170c82f.png)

jonatack commented:
Copying a wallet folder elsewhere with a different name, and attempting to load it in the GUI:
```
$ qt/bitcoin-qt -testnet
terminate called after throwing an instance of 'std::runtime_error'
  what():  BerkeleyBatch: Can't open database wallet.dat (duplicates fileid c4003e0100fe0000fd9fb0c2a56b000000000000 from wallet.dat)
Aborted
```

Screenshot of a wallet successfully loaded after I moved it from ~/.bitcoin to ~/ with an ampersand (&) displayed correctly:

![Screenshot from 2019-09-19 12-41-10](https://user-images.githubusercontent.com/2415484/65237969-342b9b00-dacb-11e9-9bfb-9644cc43f237.png)

promag commented:
@jonatack regarding the "duplicates fileid " crash see https://github.com/bitcoin/bitcoin/issues/16776#issuecomment-528672904

promag commented:
@jonatack just to be sure, if you repeat the above tests with `loadwallet` in the console you would get the same problems?

jonasschnelli commented:
Tested a bit.
What I think is hard to understand is that one needs to select a folder rather than a wallet.dat file.

I think either we allow to select wallet.dat files or inform the user that only wallet-directories are allowed ("select a wallet folder").

promag commented:
> select a wallet folder

I like that.

Sjors commented:
If you make `OpenExternalWalletActivity` a subclass of `OpenWalletActivity` it would save duplicate code, which is especially nice in `Open[External]WalletActivity::open()` with its `QTimer::singleShot`.

jonasschnelli commented:
Needs addressing comments or should otherwise be closed.

promag commented:
Will rebase and open in gui repo.

promag commented:
> Tested a bit.
> What I think is hard to understand is that one needs to select a folder rather than a wallet.dat file.
> 
> I think either we allow to select wallet.dat files or inform the user that only wallet-directories are allowed ("select a wallet folder").

I've tried to support selecting files and folders but this kind of sucks because in order be intuitive it needs non native file dialog - it allows to set a proxy model.

Anyway, I think the best approach, and considering sqlite wallets, selecting wallet.dat is indeed the best option for the moment.
